src/topology/bases.lean

```diff
@@ -10,6 +10,70 @@ import topology.order data.set.countable
 
 open set filter lattice classical
 
+namespace filter
+universe u
+variables {α : Type u}
+
+/--
+A filter has a countable basis iff it is generated by a countable collection
+of subsets of α. (A filter is a generated by a collection of sets iff it is
+the infimum of the principal filters.)
+
+Note: we do not require the collection to be closed under finite intersections.
+-/
+def has_countable_basis (f : filter α) : Prop :=
+∃ s : set (set α), countable s ∧ f = ⨅ t ∈ s, principal t
+
+lemma has_countable_basis_of_seq (f : filter α) (x : ℕ → set α) (h : f = ⨅ i, principal (x i)) :
+  f.has_countable_basis :=
+⟨range x, countable_range _, by rwa infi_range⟩
+
+lemma seq_of_has_countable_basis (f : filter α) (cblb : f.has_countable_basis) :
+    ∃ x : ℕ → set α, f = ⨅ i, principal (x i) :=
+begin
+  rcases cblb with ⟨B, Bcbl, gen⟩, subst gen,
+  classical, by_cases Bnonempty : B = ∅, { use λ n, set.univ, simp [principal_univ, *] },
+  rw countable_iff_exists_surjective_to_subtype Bnonempty at Bcbl,
+  rcases Bcbl with ⟨g, gsurj⟩,
+  rw lattice.infi_subtype',
+  use (λ n, g n), apply le_antisymm; rw le_infi_iff,
+  { intro i, apply infi_le_of_le (g i) _, apply le_refl _ },
+  { intros a, rcases gsurj a with i, apply infi_le_of_le i _, subst h, apply le_refl _ }
```

Comment:
Another mathlib style point: we generally prefer proofs like this to have one tactic per line (although this isn't strictly enforced). It makes it easier to step through the proof in an editor.